[DTensor] unbacked matmuls for no-redistribute case

pianpwk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168052
* __->__ #168051

This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78 @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

pianpwk committed:
basic unbacked matmuls

[ghstack-poisoned]

pianpwk committed:
Update on "basic unbacked matmuls"




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

- `mark_unbacked()` for DTensors allocates unbacked symbols for both inner & outer sizes (mostly for testing right now). The unbacked binding search also now matches on DTensor outer sizes (previously only inner).
- Min cost strategy selection in shard prop is a bit more unbacked friendly, choosing negative or zero-cost strategies when they're available. The case when they aren't requires comparing unbacked cost expressions, and is in the next PR (uses fallback hints).
- We assume unbacked matmul inputs are shardable by default, allowing uneven sharding/zero-size shards at runtime, if size < num shards. This alternative default path seems to be not allowing sharding on unbacked sizes at all, or DDE-ing and requiring the user to add `torch._check(size >= num shards)`. This also follows the discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068. Maybe it's also safe to extend this beyond matmuls?


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

- `mark_unbacked()` for DTensors allocates unbacked symbols for both inner & outer sizes (mostly for testing right now). The unbacked binding search also now matches on DTensor outer sizes (previously only inner).
- Min cost strategy selection in shard prop is a bit more unbacked friendly, choosing negative or zero-cost strategies when they're available. The case when they aren't requires comparing unbacked cost expressions, and is in the next PR (uses fallback hints).
- We assume unbacked matmul inputs are shardable by default, allowing uneven sharding/zero-size shards at runtime, if size < num shards. This alternative default path seems to be not allowing sharding on unbacked sizes at all, or DDE-ing and requiring the user to add `torch._check(size >= num shards)`. This also follows the discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068. Maybe it's also safe to extend this beyond matmuls?


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

laithsakka commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -464,6 +465,45 @@ def g(x):
+            torch._dynamo.reset()
+            cnt = torch._dynamo.testing.CompileCounterWithBackend("eager")
+            fn = torch.compile(torch.mm, backend=cnt, fullgraph=True)
+            fn(x_dt, y_dt)
```
shall we print results placements and tests against them to better understand the reults?

ezyang replied:
yes please

ezyang commented on torch/distributed/tensor/_ops/_matrix_ops.py:
```diff
@@ -85,8 +85,10 @@ def _mm_like_strategy(
+        if is_tensor_shardable(
+            self_strategy.shape, self_spec, allow_unbacked_sharding=True
+        ) and is_tensor_shardable(
+            mat2_strategy.shape, mat2_spec, allow_unbacked_sharding=True
```
Re

> For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the is_tensor_shardable check on the given DeviceMesh. In eager, this filters out cases where size of sharded dim < num shards. In the compiled & unbacked case, we'll often encounter dim size u_ where u_ can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: torch._check(size of sharded dim < or >= num shards), or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.

Sooooo do zero size shards work or not LOL

ezyang commented on test/distributed/tensor/test_dtensor_export.py:
```diff
@@ -552,6 +552,32 @@ def forward(self, x):
+                # 2 outer sizes, 2 inner sizes
+                self.assertEqual(len(bindings), 4)
+                n += 1
+        self.assertEqual(n, 2)  # 2 nodes with bindings (x, y)
```
test the zero size case?

ezyang commented on test/distributed/tensor/test_dtensor_export.py:
```diff
@@ -552,6 +552,32 @@ def forward(self, x):
     return (getitem,)""",  # noqa: B950
         )
 
+    def test_dtensor_mark_unbacked(self):
```
@dzmitry-huba how come this test file doesn't have LocalTensor

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

laithsakka commented:
mm
so when we create a DTensor with replicate /shared placements we probably  want to know the relation between the unbacked symbols (the locals and the globals) for example. 
```
def main():
    # Initialize process group
    dist.init_process_group(backend="nccl")

    rank = dist.get_rank()
    world_size = dist.get_world_size()
    assert world_size == 4
    device_mesh = DeviceMesh("cuda", [0,1,2,3])

    # create dummpy inputs sharedd
    value = rank + 1

    # change 1.
    local_input = torch.tensor([
        [1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.],
        [4., 4., 4.],
    ], device="cuda")  # SAME on all ranks!

    input_dtensor = DTensor.from_local(
        local_input,
        device_mesh=device_mesh,
        placements=[Replicate()],
        )

    @torch.compile(backend="inductor", fullgraph=True)
    def func(input):
        if input.size()[0] == input.to_local().size()[0]:
            return 1
        return input.size()[0]*10, input.to_local().size()[0]*10,  input*10

    torch._dynamo.decorators.mark_unbacked(input_dtensor, 0)

    func(input_dtensor)
    dist.destroy_process_group()
```
should not DDE. 

we can do this in two ways
1. ensure we do have torch checks that gets evaluate and create those relationships.
2. Similar how to how strides are assigned we can even avoid allocating new unbacked for locals and always infer from global.

laithsakka commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -158,8 +158,21 @@ def prod(xs: Iterable[int]) -> int:
+    spec: DTensorSpec,
+    allow_unbacked_sharding: Optional[bool] = None,
+) -> bool:
     """Check if the shape is shardable according to the spec."""
```
update the comment explaining how None/True/False behave differently for allow_unbacked_sharding

laithsakka commented on torch/_dynamo/decorators.py:
```diff
@@ -575,34 +575,47 @@ def mark_unbacked(
-    assert not is_traceable_wrapper_subclass(t), "not implemented yet"
+    if isinstance(t, torch.distributed.tensor.DTensor):
+        # apply on inner tensor sizes/strides
+        _apply_func_to_inner_tensors_of_same_dim(mark_unbacked, t, index)
```
we only need to handle _local_tensor its more clear to just do 
mark_unbacked(t._local_tensor, index)

laithsakka commented:
for the last point
"Lastly, testing traced redistribution decisions required using aot_eager backend, ..." you explained the issue, in the summary can you explain what change you did to address that if any?

laithsakka commented:
> did you intended to cover 1D meshes as well? this fail with DDE
> 
> ```
> 
> def main():
>     # Initialize process group
>     dist.init_process_group(backend="nccl")
> 
>     rank = dist.get_rank()
>     world_size = dist.get_world_size()
>     assert world_size == 4
>     device_mesh = DeviceMesh("cuda", [0,1,2,3])
> 
>     # create dummpy inputs sharedd
>     value = rank + 1
> 
>     from torch.distributed.tensor import rand
>     input_dtensor1 =  rand(
>             3,
>             3,
>             device_mesh=device_mesh,
>             placements=[Shard(0)],
>     )
> 
>     input_dtensor2 = rand(
>             3,
>             3,
>             device_mesh=device_mesh,
>             placements=[Shard(0)],
>         )
> 
>     @torch.compile(backend="inductor", fullgraph=True)
>     def func(i1, i2):
>         return i1@i2
> 
>     torch._dynamo.decorators.mark_unbacked(input_dtensor1, 0)
>     torch._dynamo.decorators.mark_unbacked(input_dtensor1, 1)
>     torch._dynamo.decorators.mark_unbacked(input_dtensor2, 0)
>     torch._dynamo.decorators.mark_unbacked(input_dtensor2, 1)
>     x = func(input_dtensor1, input_dtensor2)
> 
>     dist.destroy_process_group()
> ```
> 
> tlp torchrun --nproc_per_node=4 --master-port 29544 /home/lsakka/vllm/dtensor_example.py
> 
> ```
> [rank3]: torch._dynamo.exc.TorchRuntimeError: Dynamo failed to run FX node with fake tensors: call_function <built-in function matmul>(*(DTensor(local_tensor=FakeTensor(..., device='cuda:3', size=(u2, u3)), device_mesh=DeviceMesh((4,), 'cuda', stride=(1,)), placements=(Shard(dim=0),)), DTensor(local_tensor=FakeTensor(..., device='cuda:3', size=(u6, u7)), device_mesh=DeviceMesh((4,), 'cuda', stride=(1,)), placements=(Shard(dim=0),))), **{}): got RuntimeError('Could not guard on data-dependent expression 1000000.0*(FloatTrueDiv(FloatTrueDiv(3.0*(FloatTrueDiv(FloatTrueDiv(FloatTrueDiv(IntTrueDiv(4*u0*u4, 4), 1024.0), 1024.0), 1024.0)), 4.0), 87.7)) + 9.4 < 1000000.0*(FloatTrueDiv(FloatTrueDiv(12.0*(FloatTrueDiv(FloatTrueDiv(FloatTrueDiv(IntTrueDiv(4*u0*u4, 4), 1024.0), 1024.0), 1024.0)), 4.0), 87.7)) + 1000000.0*(FloatTrueDiv(FloatTrueDiv(12.0*(FloatTrueDiv(FloatTrueDiv(FloatTrueDiv(IntTrueDiv(4*u4*u5, 4), 1024.0), 1024.0), 1024.0)), 4.0), 87.7)) + 16.8 (unhinted: 1000000.0*(FloatTrueDiv(FloatTrueDiv(3.0*(FloatTrueDiv(FloatTrueDiv(FloatTrueDiv(IntTrueDiv(4*u0*u4, 4), 1024.0), 1024.0), 1024.0)), 4.0), 87.7)) + 9.4 < 1000000.0*(FloatTrueDiv(FloatTrueDiv(12.0*(FloatTrueDiv(FloatTrueDiv(FloatTrueDiv(IntTrueDiv(4*u0*u4, 4), 1024.0), 1024.0), 1024.0)), 4.0), 87.7)) + 1000000.0*(FloatTrueDiv(FloatTrueDiv(12.0*(FloatTrueDiv(FloatTrueDiv(FloatTrueDiv(IntTrueDiv(4*u4*u5, 4), 1024.0), 1024.0), 1024.0)), 4.0), 87.7)) + 16.8).  (Size-like symbols: u5, u4, u0)\n\nconsider using data-dependent friendly APIs such as guard_or_false, guard_or_true and statically_known_true.\nCaused by: return i1@i2  # vllm/dtensor_example.py:316 in func (distributed/tensor/_sharding_prop.py:638 in _select_strategy)\nFor more information, run with TORCH_LOGS="dynamic"\nFor extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u5,u4,u0"\nIf you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1\nFor more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing\n\nUser Stack (most recent call last):\n  (snipped, see stack below for prefix)\n  File "/home/lsakka/vllm/dtensor_example.py", line 316, in func\n    return i1@i2\n\nFor C++ stack trace, run with TORCHDYNAMO_EXTENDED_DEBUG_CPP=1\n\nSharding propagation failed for aten.mm.default(Spec(f32[u0, u4](S(0))), Spec(f32[u4, u5](S(0)))) on DeviceMesh((4,), \'cuda\', stride=(1,)))')
> ```
> 
> shall we add tests for 1D mesh.

same for 
```
    input_dtensor1 =  rand(
            3,
            3,
            device_mesh=device_mesh,
            placements=[Replicate() ],
    )

    input_dtensor2 = rand(
            3,
            3,
            device_mesh=device_mesh,
            placements=[Shard(0) ],
        )
 ```
 but on different location
 ```
 [rank3]: Traceback (most recent call last):
[rank3]:   File "/home/lsakka/vllm/dtensor_example.py", line 327, in <module>
[rank3]:     main()
[rank3]:   File "/home/lsakka/vllm/dtensor_example.py", line 322, in main
[rank3]:     x = func(input_dtensor1, input_dtensor2)
[rank3]:   File "/home/lsakka/pytorch10/pytorch/torch/_dynamo/eval_frame.py", line 930, in compile_wrapper
[rank3]:     raise e.with_traceback(None) from e.__cause__  # User compiler error
[rank3]: torch._dynamo.exc.UserError: Could not guard on data-dependent expression Eq(((u3 + 3)//4), 0) (unhinted: Eq(((u3 + 3)//4), 0)).  (Size-like symbols: u3)

[rank3]: consider using data-dependent friendly APIs such as guard_or_false, guard_or_true and statically_known_true.
[rank3]: Caused by: return i1@i2  # vllm/dtensor_example.py:316 in func (_ops.py:881 in decompose)
[rank3]: For more information, run with TORCH_LOGS="dynamic"
[rank3]: For extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u3"
[rank3]: If you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
[rank3]: For more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing

[rank3]: User Stack (most recent call last):
[rank3]:   (snipped, see stack below for prefix)
[rank3]:   File "/home/lsakka/vllm/dtensor_example.py", line 316, in func
[rank3]:     return i1@i2

[rank3]: For C++ stack trace, run with TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
[rank3]: For more information about this error, see: https://pytorch.org/docs/main/generated/exportdb/index.html#constrain-as-size-example

[rank3]: from user code:
```

or is it the case tha there is no **zero-cost strategies** for those so they are not handled in this PR?

laithsakka commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -172,7 +185,7 @@ def is_tensor_shardable(shape: Sequence[int], spec: DTensorSpec) -> bool:
         # TODO: maybe we should determine is_shardable based on
         #       whether it's evenly sharded or not
-        if shards_map[i] > 1 and dim_size < shards_map[i]:
+        if shards_map[i] > 1 and guard_fn(dim_size < shards_map[i]):
```
question what that is the value of this filtering in eager? or no unbacked case?
i guess we are saying if we know for sure this strategy will end up leaving some ranks with 0 size tensors. then probably its not useful lets exclude it?

probably the function name ""is_tensor_shardable"" is not very suitable
or is that case that if we really end up sending tensor of size 3 to 4 ranks we consider this not shardable situation? and have this assumption in other places?

pianpwk replied:
Hmm one case is if you have a matmul x @ y with statically known shapes [1, 8] x [8, 1], it's a bit of a waste to shard x.size(0) or y.size(1) on a device mesh.

> probably the function name ""is_tensor_shardable"" is not very suitable

I think I agree... at least once the dynamic shape case is considered.

laithsakka replied:
I meant it should be called is_tensor_shardable, because it does not mean we only execlude non shardable cases its more like , is_it_worth_sharding(.._

laithsakka commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -158,8 +158,21 @@ def prod(xs: Iterable[int]) -> int:
+def is_tensor_shardable(
+    shape: Sequence[int],
+    spec: DTensorSpec,
+    allow_unbacked_sharding: Optional[bool] = None,
```
we never use allow_unbacked_sharding=False? do we

laithsakka replied:
I also see all usages of is_tensor_shardable using allow_unbacked_sharding = True
except the one in expand_to_full_mesh_op_strategy

pianpwk replied:
yeah expand_to_full_mesh_op_strategy is very generally used

pianpwk commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -464,6 +458,54 @@ def g(x):
+            out = fn(x_dt, y_dt)
+
+            # test sharded matmuls with zero-size shards
+            if test_index == 1:
```
just expensive (i'll add more tests in the follow-up PR), this is also the only one that shards on the 0-size dim

laithsakka replied:
i mean its a unit test i would just remove line   if test_index == 1: but to u lol

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

laithsakka commented on torch/_dynamo/decorators.py:
```diff
@@ -575,34 +575,47 @@ def mark_unbacked(
+    if isinstance(t, torch.distributed.tensor.DTensor):
+        # apply on inner tensor sizes/strides
+        _apply_func_to_inner_tensors_of_same_dim(mark_unbacked, t, index)
+    elif is_traceable_wrapper_subclass(t):
```
else
   assert not X?

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78 H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] unbacked matmuls for no-redistribute case"


This allows compiling a matmul on 2 DTensors with fully unbacked sizes, when a zero-cost strategy is available.

Changes with the PR:
- `mark_unbacked()` would previously error on tensor subclasses; now for DTensors it allocates unbacked symbols for both inner & outer sizes. The main motivation here is for testing, so happy to tweak semantics. The unbacked binding search process also now matches on DTensor outer sizes.
- Selecting an op strategy in sharding propagation is based on minimal redistribution costs, and these costs are functions of tensor shapes, so can be unbacked expressions. This PR makes this process more unbacked-friendly, choosing negative or zero-cost strategies when they're available. When these "trivial" strategies aren't available, selection requires comparing unbacked costs, addressed in the next PR (with usage of fallback hints).
- For matmul strategies, sharding prop rules filter out strategies where the matmul inputs fail the `is_tensor_shardable` check on the given DeviceMesh. In eager, this filters out cases where `size of sharded dim < num shards`. In the compiled & unbacked case, we'll often encounter dim size `u_` where `u_` can be both larger and smaller than num shards. This PR assumes such cases are shardable by default, and the implication is that strategies that shard on unbacked dimensions are included for consideration, and if selected, can lead to uneven sharding/zero-size shards at runtime. Alternatives would be 1) the current state of things: DDE and force the user to pick a path: `torch._check(size of sharded dim < or >= num shards)`, or 2) assume the non-shardable case and never include sharded strategies, unless the user picks the shardable path. More discussion in https://github.com/pytorch/pytorch/issues/165034#issuecomment-3417695068.
- Lastly, testing traced redistribution decisions required using aot_eager backend, so that the collectives/ops were hardcoded (eager backend would go through DTensor.dispatch again). This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78 H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
